collectors/namespace.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2017 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collectors
+
+import (
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/net/context"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	descNamespaceLabelsName          = "kube_namespace_labels"
+	descNamespaceLabelsHelp          = "Kubernetes labels converted to Prometheus labels."
+	descNamespaceLabelsDefaultLabels = []string{"namespace"}
+
+	descNamespaceCreated = prometheus.NewDesc(
+		"kube_namespace_created",
+		"Unix creation timestamp",
+		[]string{"namespace"}, nil,
+	)
+
+	descNamespaceLabels = prometheus.NewDesc(
+		descNamespaceLabelsName,
+		descNamespaceLabelsHelp,
+		descNamespaceLabelsDefaultLabels, nil,
+	)
+
+	descNamespacePhase = prometheus.NewDesc(
+		"kube_namespace_status_phase",
+		"kubernetes namespace status phase.",
+		[]string{
+			"name",
+			"phase",
+		}, nil,
+	)
+)
+
+// NamespaceLister define NamespaceLister type
+type NamespaceLister func() ([]v1.Namespace, error)
+
+// List return namespace list
+func (l NamespaceLister) List() ([]v1.Namespace, error) {
+	return l()
+}
+
+// RegisterNamespaceCollector registry namespace collector
+func RegisterNamespaceCollector(registry prometheus.Registerer, kubeClient kubernetes.Interface, namespace string) {
+	client := kubeClient.CoreV1().RESTClient()
+	var selector fields.Selector
+	if namespace != api.NamespaceAll {
```

Comment:
it's safe to query for all namespaces here, the namespace restriction is only meant for limiting RBAC roles